---
title: "Mapping report of BD Rhapsody targeted data"
date: "`r Sys.Date()`"
author: "ImmunoGeneTeqs Inc."
output:
  rmdformats::readthedown:
    highlight: kate
    self_contained: true
    thumbnails: false
    lightbox: false
    gallery: false
    mathjax: false
    toc_depth: 2  #type-of-column header specification 
    md_extensions: -ascii_identifiers
    css: styles.css
params:
  samplename: samplename
  species: species
  umi: umi
  expectCells: expectCells
---

```{r setup, echo=FALSE, cache=FALSE, message=FALSE}
suppressMessages(library(knitr))
suppressMessages(library(rmdformats))
suppressMessages(library(data.table))
suppressMessages(library(Seurat))
suppressMessages(library(dplyr))
suppressMessages(library(ggplot2))
suppressMessages(library(stringr))
suppressMessages(library(tibble))
suppressMessages(library(tidyr))
suppressMessages(library(DropletUtils))
suppressMessages(library(DT))
suppressMessages(library(cowplot))
suppressMessages(library(Matrix))
reticulate::use_python(
  python = "/usr/bin/python3",
  required = TRUE
)
library(reticulate)
py_set_seed(42, disable_hash_randomization = TRUE)

quiet <- function(x) { 
  sink(tempfile()) 
  on.exit(sink()) 
  invisible(force(x)) 
} 

tableread_fast = function(i){
	 tmp = fread(i, header=FALSE, sep="\t", quote="")
   tmp = as.data.frame(tmp)
	 return(tmp)
}

## Global options
options(max.print="75")
opts_chunk$set(echo=FALSE,
	             cache=TRUE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE)
opts_knit$set(width=50)
opts_template$set(
  fig_small = list(fig.width=5, fig.height=4, dpi=300, out.width = 200, out.height = 160), 
  fig_large = list(fig.width=5, fig.height=4, dpi=300, out.width = 500, out.height = 400)
  )

sample_name = params$samplename
umi = as.logical(params$umi)
expectCells = as.numeric(params$expectCells)

if(params$species == "hsa"){
  reference = "GRCh38 release-107"
    GENCODE = "gencode.v41"
} else if (params$species == "mmu"){
  reference = "GRCm39 release-107"
  GENCODE = "gencode.vM30"
} else if (params$species == "macaca"){
  reference = "Macaca_fascicularis_6.0 release-107"
      GENCODE = "ensembl release-107"
} else if (params$species == "rat"){
  reference = "Rattus_norvegicus.Rnor_6.0 release-107"
  GENCODE = "ensembl release-107"
} else if (params$species == "mermoset"){
  reference = "Callithrix_jacchus.ASM275486v1 release-104"
}
```

--------

## Base composition of adapter-trimmed/quality-filtered sequencing data

After adapter trimming and quality filtering, base composition of sequencing data was analyzed by FastQC-v0.11.9.  

Figure 1a. Structure of read1 (cell barcode) of BD Rhapsody Enhanced beads.  

```{r, echo=FALSE, out.width = 500, out.height = 200, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url1 = paste("../reference/BD_CB/CB_structure.png", sep="")
knitr::include_graphics(paste("../reference/BD_CB/CB_structure.png", sep=""))
```

Figure 1b. Base composition of read1 data (cell barcodes).   

```{r, echo=FALSE, out.width = 600, out.height = 250, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url1 = paste(sample_name, "_per_base_sequence_content_R1.png", sep="")
knitr::include_graphics(paste("../data/", sample_name,"_per_base_sequence_content_R1.png", sep=""))
```

Figure 1b. Base composition of read2 data (cDNA sequences). Note that the first few bases did not show uniform distribution becuase the sequencing library was generated by enzymatic fragmentation.  
```{r, echo=FALSE, out.width = 800, out.height = 250, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url2 = paste(sample_name, "_per_base_sequence_content_R2.png", sep="")
knitr::include_graphics(paste("../data/", sample_name,"_per_base_sequence_content_R2.png", sep=""))
```

Base composition plot was stored at `r paste("./plots/", url1, sep="")` and `r paste("./plots/", url2, sep="")`.

--------

## Brief summary of the mapping pipeline procedures

Pair-end Fastq files (R1: cell barcode reads, R2: RNA reads) of sample `r paste(sample_name, sep="")` were processed as follows.  
Adapter trimming, quality filtering, and phase-shift base removal of sequencing data was performed by using Cutadapt 4.1.  
Filtered cell barcode and cDNA reads were annotated and mapped to reference genome (build `r paste(reference,sep="")`) by using STARsolo (2.7.10a) and known BD Rhapsody cell barcodes by the following parameters:--clipAdapterType CellRanger4 --outSAMmultNmax 1 --outFilterScoreMinOverLread 0 --outFilterMatchNminOverLread 0 --outFilterMultimapScoreRange 0 --seedSearchStartLmax 30 --soloType CB_UMI_Complex --soloUMIdedup `r if(umi){paste("Exact")}else{paste("NoDedup")}` --soloMultiMappers Rescue --soloFeatures Gene GeneFull --soloCBmatchWLtype EditDist_2 --soloUMIlen 8 --soloAdapterSequence NNNNNNNNNGTGANNNNNNNNNGACA --soloCBposition 2_0_2_8 2_13_2_21 3_1_3_9 --soloUMIposition 3_10_3_17.   
`r if(umi){paste("Raw read counts were compressed by 8-base BD Rhapsody beads.")}else{paste("We do not perform UMI compression because BD Rhapsody UMI length (8-base and directly attatched polyT) is not sufficient to avoid UMI collision (minimum 10-base is required, Perukhov et al Genome Biol 2018).")}`   
Resulted `r if(umi){paste("UMI")}else{paste("non-umi")}` count data was converted to genes x cells matrix file and inflection threshold of the barcode rank plot was detected by using DropletUtils package in `r paste(version$version.string, sep="")`.  
Valid cell barcodes and backgrounds from inflection threshold * 0.9 to inflection threshold * 1.1 were separated by Dropkick package (Heiser et al Genome Biol 2022).  
  
For STAR index build, small RNA annotations were filtered out from `r paste(GENCODE)` gtf file.  
Spliced/unspliced count matrix files for RNA velocity analysis was created by using STARsolo Gene (spliced count data) and GeneFull (spliced+unspliced count data) matrixes. Unspliced count matrix was generated by (GeneFull count matrix) - (Gene count matrix).  

```{r analysis, echo=FALSE, cache=FALSE, echo=FALSE}

#create output folders
setwd("../") #must set working directory. default is .Rmd-existing path.

#make output directories

#make output directories
dir.name = sprintf("./result/%s_results", sample_name)
if(!dir.exists(dir.name)){
dir.create(dir.name)
}
dir.name.1 = paste(dir.name, "/matrix", sep="")
if(!dir.exists(dir.name.1)){
dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/plots", sep="")
if(!dir.exists(dir.name.1)){
dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/stats", sep="")
if(!dir.exists(dir.name.1)){
 dir.create(dir.name.1)
}
dir.name.1 = paste(dir.name, "/processed", sep="")
if(!dir.exists(dir.name.1)){
 dir.create(dir.name.1)
}

#read count table by using Read10X function
res4 = Read10X(paste("./result/", params$samplename,"_results/processed", sep=""))

#remove "_" from cell barcodes
res4 = res4[Matrix::rowSums(res4)>0,]

# calculate Total read counts and Detected Genes of each Cell-BCs by data.table package
res2 = CreateSeuratObject(res4, min.cells=1, min.features=1)
res2 = cbind(rownames(res2@meta.data), res2@meta.data[,c(2,3)])
colnames(res2) = c("ALLCell_BCs", "Total_reads", "Symbol_kinds")
res2 = res2[order(res2$Total_reads, decreasing = T),]
quiet(gc())

#reorder count matrix by total count

res4 = res4[,order(Matrix::colSums(res4), decreasing=T)]
symbol_total_200000 = nrow(res4)

#detect inflection point/knee-threshold.

if(expectCells >= 10000){
upper_limit = expectCells * 0.01
} else if (expectCells >= 5000){
upper_limit = expectCells * 0.02
} else if (expectCells >= 1000){
upper_limit = 50  
} else {
upper_limit = 10
}
br.out = barcodeRanks(res4[,c(upper_limit:100000)])



#define threshold as inflection point of knee-plot
tmp = res4[,Matrix::colSums(res4)>=metadata(br.out)$inflection] 
tmp1 = c(ncol(tmp), 10000)

tmp_knee = res4[,Matrix::colSums(res4)>=metadata(br.out)$knee]

inflection_threshold=ncol(tmp)
knee_threshold = ncol(tmp_knee)

#correct erroneous inflection point detection if inflection threshold > knee threshold

if (knee_threshold>=inflection_threshold){
br.out = barcodeRanks(res4[,c(knee_threshold:100000)])
tmp = res4[,Matrix::colSums(res4)>=metadata(br.out)$inflection] 
tmp1 = c(ncol(tmp), 10000)

tmp_knee = res4[,Matrix::colSums(res4)>=metadata(br.out)$knee]

inflection_threshold=ncol(tmp)
knee_threshold = ncol(tmp_knee)
}

tmp2 = c(ncol(tmp), metadata(br.out)$inflection)
symbol_total = nrow(tmp)

tmp = res4[,c(1:floor(inflection_threshold*1.1))]

cellbarcode_ambiguous = colnames(res4[,c(floor(1+(inflection_threshold*0.9)):floor(inflection_threshold*1.1))])
cellbarcode = colnames(res4[,c(1:floor(inflection_threshold*0.9))])

#read dropkick estimation result
dropkick_label = fread(paste0("./result/", sample_name, "_results/processed/", sample_name, "_counts_top80000BC_dropkick_scores.csv.gz"), 
                                       sep = ",", header = T)
dropkick_label = as.data.frame(dropkick_label)
dropkick_label = dropkick_label[dropkick_label$dropkick_label == TRUE,]

#export valid cell barcodes from knee point to inflection point*1.1

valid_cell = c(cellbarcode, cellbarcode_ambiguous[cellbarcode_ambiguous %in% dropkick_label$V1])
tmp = res4[,valid_cell] 


#create loom file for RNA velocity analysis
res4_spliced = Read10X(paste("./result/", params$samplename,"_results/Gene", sep=""))
colnames(res4_spliced)=gsub("_", "", colnames(res4_spliced))

res4_unspliced = Read10X(paste("./result/", params$samplename,"_results/processed", sep=""))
colnames(res4_unspliced)=gsub("_", "", colnames(res4_unspliced))


Nbarcode = ncol(res4_unspliced)

res4_spliced = res4_spliced[,as.character(valid_cell)] 
res4_unspliced = res4_unspliced[,as.character(valid_cell)] 
res4_full = res4_unspliced
res4_unspliced = res4_unspliced - res4_spliced

#export spliced and unspliced count matrix
save.folder=paste0("./result/", params$samplename, "_results/velocyto/spliced/")
write10xCounts(path=save.folder, x=res4_spliced, barcodes=colnames(res4_spliced), gene.id = rownames(res4_spliced),
               gene.symbol = rownames(res4_spliced), overwrite=TRUE, type="sparse", version="3")

save.folder=paste0("./result/", params$samplename, "_results/velocyto/unspliced/")
write10xCounts(path=save.folder, x=res4_unspliced, barcodes=colnames(res4_unspliced), gene.id = rownames(res4_unspliced),
               gene.symbol = rownames(res4_unspliced), overwrite=TRUE, type="sparse", version="3")

#export to file
file.name_matrix = sprintf("matrix_inflection_%s.txt.gz", sample_name) 
file.name_matrix = paste(dir.name, "/matrix/", file.name_matrix, sep="")
fwrite(as.data.frame(tmp), file.name_matrix, row.names=T, col.names=T, sep="\t", 
       eol="\n", quote=F, nThread=32, compress="gzip")

tmp2 = c(ncol(tmp), min(colSums(tmp)))
symbol_total = nrow(tmp)

#export thresholds and survived cells
threshold_log = cbind(tmp1, tmp2)
rownames(threshold_log)=c("Survived_Cells", "count_threshold")
colnames(threshold_log)=c("count", "inflection")

#concatenate and reshape mapping statistics
res2=as.data.frame(res2)
res3 = res2[res2$ALLCell_BCs %in% valid_cell,]
read_stats = c(median(res3[,2]), mean(res3[,2]), sum(res3[,2]), sum(res2[,2]))
symbol_stats =c(median(res3[,3]), mean(res3[,3]), symbol_total, symbol_total_200000)
inflection_log = cbind(read_stats, symbol_stats)
rownames(inflection_log)=c("median", "mean", "total_inflection", "total")

cutadapt_log = sprintf("./result/%s_cutadapt.log", sample_name)
cutadapt_log = read.table(cutadapt_log, header=TRUE, row.names=1, sep="\t", 
                         quote="",stringsAsFactors=FALSE)

#read STARsolo mapping result
STARsolo_log = paste0("./result/", sample_name, "_results/stats/STARsolo_Summary.csv")
STARsolo_log = read.csv(STARsolo_log, header=T, 
                        quote="", stringsAsFactors = F)

cutadapt_log1 = cutadapt_log[,1]
mapping_log1 = STARsolo_log[c(1,2,6,7,8),2]
threshold_log1 = threshold_log[1,1]
threshold_log2 = threshold_log[c(2,1),2]
inflection_log1 = inflection_log[c(3,1,2),1]
inflection_log2 = inflection_log[c(1:3),2]
mapping_log3 = round(inflection_log[3,1]/inflection_log[4,1]*100, digits=2)

logs = c(cutadapt_log1, 
         mapping_log1[1], 
         round(mapping_log1[2]*100, digits=2),
         round(mapping_log1[3]*100, digits=2),
         round(mapping_log1[4]*100, digits=2),
         round(mapping_log1[5]*100, digits=2),
         inflection_log[4,1], 
         threshold_log2, 
         threshold_log1, 
         inflection_log1[1], 
         mapping_log3, 
         inflection_log1[2:3], 
         inflection_log2)
logs = as.data.frame(logs)
rownames(logs) = c("Total raw reads", 
                   "Total reads after trimming and quality filtering", 
                   "Cell barcode calling rate", 
                   "Mapping rate against genome", 
                   "Unique-mapping rate against genome", 
                   "Feature assignment rate", 
                   "Final usable reads",
                   "Minimum read number of survived cells", 
                   "Survived cells", 
                   "over10000read cells",
                   "Total reads of valid cells", 
                   "Percent of total reads of valid cells",
                   "Median reads per cell", 
                   "Mean reads per cell", 
                   "Median symbols per cell", 
                   "Mean symbols per cell", 
                   "Total genes detected")
colnames(logs) = sample_name
logs=logs[c(1:9,11:17),,drop=F]
logs[c(3:6,11),1] = round(logs[c(3:6,11),1], digits=2)
logs[c(1,2,7:10,12:16),1] = round(logs[c(1,2,7:10,12:16),1], digits=0)
file.name.logs = sprintf("stats_concat_%s.txt", sample_name) 
file.name.logs = paste(dir.name, "/stats/", file.name.logs, sep="")
write.table(logs, file.name.logs, row.names=T, col.names=T, sep="\t", quote=F, append=F)

logs[,1]=as.character(logs[,1])

#export plots

df  = res2[1:min(1000000, Nbarcode),]  
df$Rank = c(1:min(1000000, Nbarcode))
colnames(df)[2] = "Nreads"
df[,2]=log10(df[,2])
df[,4]=log10(df[,4])


df1_1 = df[c(1:floor(inflection_threshold*0.9)),] 
df1_2 = df[(floor(inflection_threshold*1.1)+1):min(1000000, Nbarcode),] 
df1_3 = df[c(floor(1+(inflection_threshold*0.9)):floor(inflection_threshold*1.1)),]
df1_1$group = rep("Cell", nrow(df1_1))
df1_2$group = rep("Background", nrow(df1_2))
df1_3$group = rep("Cell or background", nrow(df1_3))
df = rbind(df1_1, df1_3, df1_2)

df$group = factor(df$group, levels = c("Cell", "Cell or background", "Background"))


p_knee = ggplot(df) + 
  geom_point(data=df, aes(Rank, Nreads, group=group, color=group), size=0.5) + 
  scale_color_manual(values=c("#006699", "#add8e6", "#d3d3d3")) +
  scale_fill_manual(values=c("#006699","#add8e6", "#d3d3d3")) +
  geom_line(data = df1_1, aes(Rank, Nreads), colour="#006699", size=1) +
  geom_line(data = df1_3, aes(Rank, Nreads), colour="#add8e6", size=1) +
  geom_line(data = df1_2, aes(Rank, Nreads), colour= "#d3d3d3", size=1) +
  ylab(expression(paste("Read number (", {log[10]}, ")", sep=""))) +  
  xlab(expression(paste("Cell barcode rank (", {log[10]}, ")", sep=""))) +
    scale_x_continuous(breaks=c(0,1,2,3,4,5,6))+
  ggtitle(paste("Barcode rank plot (survived cells = ", threshold_log[1,2]," cells)", sep="")) +
  theme_linedraw() +
  theme(plot.title=element_text(hjust = 0.5), text=element_text(size=10)) + 
  guides(colour = guide_legend(override.aes = list(size=5)))


#plotting counts vs Detected genes over inflection threshold
res3 = res2[res2[,2]>=metadata(br.out)$inflection,]
df = data.frame(Reads=res3[,2], 
                Symbols=res3[,3])

p_count_symbol = ggplot(df,  aes(Reads, Symbols)) +
    geom_point(data=df, aes(Reads, Symbols), size=0.5, alpha=0.5, shape=20, colour="#006699") +
    theme_linedraw() +
    ggtitle("Read counts versus symbols")+
    theme(plot.title=element_text(hjust = 0.5), 
          legend.position = 'none', text=element_text(size=10))
file.name = sprintf("CountvsSymbol_%s.png", sample_name)
file.name = paste(dir.name, "/plots/", file.name, sep="")
ggsave(file=file.name, plot=p_count_symbol, device="png", units="in",
       dpi=300, width=5.5, height=4, limitsize = FALSE)

dropkick_label = fread(paste0("./result/", sample_name, 
                              "_results/processed/", sample_name, "_counts_top80000BC_dropkick_scores.csv.gz"), 
                                       sep = ",", header = T)
dropkick_label = as.data.frame(dropkick_label)
rownames(dropkick_label)=dropkick_label$V1
dropkick_score = dropkick_label[dropkick_label$V1 %in% colnames(res4),]
dropkick_score$rank = log10(c(1:nrow(dropkick_score)))
label = dropkick_score$dropkick_label
label[label %in% TRUE]="Cell"
label[label %in% FALSE]="Background"
label[c(1:floor(inflection_threshold*0.9))]="Cell"
label[c((floor(inflection_threshold*1.1)+1):nrow(dropkick_score))]="Background"
label = factor(label)
dropkick_score$dropkick_label = label
#dropkick_score$alpha = abs(dropkick_score$dropkick_score - 0.5)
#draw valid cell probability figure
p_dropkick = ggplot(dropkick_score) + 
  geom_point(data=dropkick_score, aes(rank, dropkick_score, group=dropkick_label, 
                                      color=dropkick_label), alpha=0.5, size=0.5) +
  geom_vline(aes(xintercept=log10(floor(inflection_threshold*0.9)), color="Inflection*0.9"), size=0.5) +
  geom_vline(aes(xintercept=log10(floor(inflection_threshold*1.1)), color="Inflection*1.1"), size=0.5) +
  scale_color_manual(values=c(Background = "#dc143c", Cell = "#006699", "Inflection*0.9" = "cyan", "Inflection*1.1" = "magenta")) +
  scale_fill_manual(values=c(Background = "#dc143c", Cell = "#006699", "Inflection*0.9" = "cyan", "Inflection*1.1" = "magenta")) +
  ylab("Dropkick score") +  
  xlab(expression(paste("Cell barcode rank (", {log[10]}, ")", sep=""))) +
  scale_x_continuous(breaks=c(0,1,2,3,4,5))+
  ggtitle("Dropkick estimation plot") +
  theme_linedraw() +
  theme(plot.title=element_text(hjust = 0.5), text=element_text(size=10))+
  guides(colour = guide_legend(override.aes = list(size=5)))


file.name = sprintf("dropkick_score_%s.png", sample_name)
file.name = paste(dir.name, "/plots/", file.name, sep="")
ggsave(file=file.name, plot=p_dropkick, device="png", units="in",
       dpi=300, width=5.5, height=4, limitsize = FALSE)

```
---

## Mapping statistics

---

```{r output_table, echo=FALSE, cache=FALSE,dpi=300, out.width = 500, out.height = 500}
DT::datatable(logs, filter="none", width=700,
          caption = 'Table 1: mapping statistics, detected cell and detected gene statistics.', 
          class = 'cell-border stripe', 
          options = list(pageLength=16, searching=FALSE, lengthChange=FALSE)) %>%
   DT::formatStyle(columns = c(1, 2), fontSize = 9)
```

Original data is stored at `r sprintf("/stats/stats_concat_%s.txt", sample_name)`

---

## Barcode rank plot, Dropkick plot, and read counts/gene symbols scatter plot  {.tabset .tabset-fade}

Figure 2. Barcode rank plot ((log10 read counts of each cell barcodes) x (rank of read counts of each cell barcodes)), Dropkick score/qc plots, and scatter plot (read count vs number of detected genes for each cell).  
Higher dropkick score means high probability of valid cells.  

---

### Barcode rank plot

```{r barcode_rank_plot, echo=FALSE, cache=FALSE, fig.width=6, fig.height=4.5, dpi=450, out.width = 600, out.height = 400, fig.align='center'}
plot_grid(p_knee)
  quiet(dev.off())
  quiet(gc())
  Sys.sleep(5)
```

### Dropkick score plot  

```{r, echo=FALSE, cache=FALSE, out.width = 700, out.height = 450, fig.width=7, fig.height=4.5, fig.align='center', dpi=450}
plot_grid(p_dropkick)
  quiet(dev.off())
  quiet(gc())
  Sys.sleep(5)
```

Cyan line means knee-point of barcode rank plot. Barcodes from rank 1 to inflection point*1.1 were shown.  

### Dropkick QC plot  

```{r, echo=FALSE, out.width = 800, out.height = 450, message=FALSE, fig.align='center', dpi=300}
# Define variable containing url
url2 = paste("../result/", sample_name, "_results/plots/", sample_name, "_counts_top80000BC_qc.png", sep="")
knitr::include_graphics(url2)
```

### Scatter plot

```{r scatter_plot, echo=FALSE, cache=FALSE, fig.width=6, fig.height=4.5, dpi=450, out.width = 600, out.height = 450, fig.align='center'}
plot_grid(p_count_symbol)
    quiet(dev.off())
  quiet(gc())
  Sys.sleep(5)

```

---  

---

## Output file explanations


Raw sequence files (fastq.gz format)  

* Raw sequence fastq.gz files were stored at `r paste("./fastq/", sep="")` folder. "R1" means read 1 sequence, and "R2" means read 2 sequence.  

Gene-expression matrix files (gzip compressed)

Rownames are gene symbols, and colnames are cell barcodes.

* Genes x cells matrix (background filtered) was stored at `r sprintf("./matrix/matrix_inflection_%s.txt.gz", sample_name)`.
* Unfiltered genes x cells matrix was stored at `r paste("./processed/", sep="")` folder.  
* Spliced and unspliced count matrix files for velocyto analysis were stored at `r paste("./velocyto/", sep="")` folder.  

Plot file locations  

* Barcode rank plot was stored at `r sprintf("./plots/BarcodeRankPlot_%s.png", sample_name)`.  
* Dropkick plot was stored at `r sprintf("./plots/%s_counts_qc.png", sample_name)` and `r sprintf("./plots/dropkick_score_%s.png", sample_name)` .  
* Scatter plot was stored at `r sprintf("./plots/CountvsSymbol_%s.png", sample_name)`.  

---

## Sessioninfo

```{r sessioninfo, results.folding="hide", summary.results="SessionInfo"}  
options(max.print = 999999)
sessionInfo()
```
